src/psyclone/dynamo0p3.py

```diff
@@ -7452,137 +7451,136 @@ def gen_code(self, parent):
+
+        :raises GenerationError: if the contents of the arguments do \
+            not match.
+        :raises InternalError: if an unexpected datatype is found.
+
+        '''
+        # 1: intrinsic datatype
+        actual_datatype = kern_code_arg.datatype.intrinsic
+        expected_datatype = interface_arg.datatype.intrinsic
+        if actual_datatype != expected_datatype:
+            raise GenerationError(
+                "Kernel argument '{0}' has datatype '{1}' "
+                "in kernel '{2}' but the LFRic API expects '{3}'."
+                "".format(kern_code_arg.name, actual_datatype,
+                          self.name, expected_datatype))
+        # 2: precision
+        actual_precision = kern_code_arg.datatype.precision
+        expected_precision = interface_arg.datatype.precision
+        if actual_precision.name != expected_precision.name:
+            raise GenerationError(
+                "Kernel argument '{0}' has precision '{1}' "
+                "in kernel '{2}' but the LFRic API expects '{3}'."
+                "".format(kern_code_arg.name, actual_precision.name,
+                          self.name, expected_precision.name))
+        # 3: intent
+        actual_intent = kern_code_arg.interface.access
+        expected_intent = interface_arg.interface.access
+        if actual_intent.name != expected_intent.name:
+            raise GenerationError(
+                "Kernel argument '{0}' has intent '{1}' "
+                "in kernel '{2}' but the LFRic API expects intent '{3}'."
+                "".format(kern_code_arg.name, actual_intent.name,
+                          self.name, expected_intent.name))
+        # 4 scalar or array
+        if interface_arg.is_scalar:
+            if not kern_code_arg.is_scalar:
+                raise GenerationError(
+                    "Kernel argument '{0}' is expected to be a scalar "
+                    "by the LFRic API in kernel '{1}' but it is not."
+                    "".format(kern_code_arg.name, self.name))
+        elif interface_arg.is_array:
+            if not kern_code_arg.is_array:
+                raise GenerationError(
+                    "Kernel argument '{0}' is expected to be an array "
+                    "by the LFRic API in kernel '{1}' but it is not."
+                    "".format(kern_code_arg.name, self.name))
+            # 4.1 array arguments
+            for dim_idx, kern_code_arg_dim in enumerate(kern_code_arg.shape):
+                interface_arg_dim = interface_arg.shape[dim_idx]
+                if (isinstance(kern_code_arg_dim, DataSymbol) and
+                        isinstance(interface_arg_dim, DataSymbol)):
+                    # Only check when there is a symbol. Unspecified
+                    # dimensions, dimensions with scalar values,
+                    # offsets, or dimensions that include arithmetic
+                    # are skipped.
+                    self._validate_kernel_code_arg(
```

Comment:
Ooh, I love a bit of recursion :-) Having said that though, I think that using it here may result in confusing error messages. I *think* I'd be happy if you protected this with a `try` and then re-raised any exception with a more meaningful message.

Reply:
OK, I've modified the code. Hopefully it is to your liking. I also added a check that the number of arguments are the same.